beacon-chain/rpc/validator/proposer.go

```diff
@@ -353,6 +345,48 @@ func (vs *Server) canonicalEth1Data(ctx context.Context, beaconState *stateTrie.
 	return canonicalEth1Data, canonicalEth1DataHeight, nil
 }
 
+func (vs *Server) depositTrie(ctx context.Context, canonicalEth1DataHeight *big.Int) (*trieutil.SparseMerkleTrie, error) {
+	var depositTrie *trieutil.SparseMerkleTrie
+
+	var finalizedDeposits *depositcache.FinalizedDeposits
+	if featureconfig.Get().EnableFinalizedDepositsCache {
+		finalizedDeposits = vs.DepositFetcher.FinalizedDeposits(ctx)
```

Comment:
over here you need to fetch a copy of the finalized trie, not the actual trie stored in the cache. Otherwise you will end up mutating the trie, everytime you use this method